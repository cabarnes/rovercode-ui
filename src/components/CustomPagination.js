import React from 'react';
import { Icon, Pagination } from 'semantic-ui-react';
import { hot } from 'react-hot-loader';

const CustomPagination = (props) => (
  <Pagination
    {...props}
    ellipsisItem={{ content: <Icon name="ellipsis horizontal" />, icon: true }}
    firstItem={{ content: <Icon name="angle double left" />, icon: true }}
    lastItem={{ content: <Icon name="angle double right" />, icon: true }}
    prevItem={{ content: <Icon name="angle left" />, icon: true }}
    nextItem={{ content: <Icon name="angle right" />, icon: true }}
  />
);

export default hot(module)(CustomPagination);
