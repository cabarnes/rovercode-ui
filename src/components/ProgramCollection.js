import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import {
  Box,
  Container,
  Grid,
  Button,
  Typography,
  Chip,
  TextField,
  InputAdornment,
  Menu,
  MenuItem,
  Card,
  CardContent,
  CardHeader,
  CardActionArea,
  CardActions,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import {
  Add,
  ArrowDownward,
  ArrowUpward,
  Search,
} from '@material-ui/icons';
import { Pagination, Autocomplete } from '@material-ui/lab';
import { Link } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';

class ProgramCollection extends Component {
  constructor(props) {
    super(props);

    this.state = {
      page: 1,
      searchQuery: null,
      ordering: 'name',
      tagFilters: [],
      sortMenuAnchorElement: null,
    };
  }

  update = () => {
    const { onUpdate } = this.props;
    const {
      page,
      ordering,
      searchQuery,
      tagFilters,
    } = this.state;

    const params = {
      page,
      ordering,
      tag: tagFilters.join(),
    };

    if (searchQuery) {
      params.search = searchQuery;
    }

    onUpdate(params);
  }

  toggleOrdering = (name) => {
    const { ordering } = this.state;

    if (ordering.endsWith(name)) {
      if (ordering.startsWith('-')) {
        return ordering.substr(1);
      }

      return `-${ordering}`;
    }

    return name;
  }

  handleSortClick = (event) => this.setState({
    sortMenuAnchorElement: event.currentTarget,
  })

  handleSortClose = () => this.setState({
    sortMenuAnchorElement: null,
  })

  handlePageChange = (e, page) => this.setState({
    page,
  }, () => this.update())

  handleSearchChange = (event) => this.setState({
    searchQuery: event.target.value,
    page: 1,
  }, () => this.update())

  handleNameOrderingChange = () => this.setState({
    ordering: this.toggleOrdering('name'),
  }, () => this.update())

  handleTagFilterChange = (event, value) => this.setState({
    tagFilters: value,
  }, () => this.update())

  render() {
    const {
      user,
      label,
      onProgramClick,
      onRemoveClick,
      programs,
      intl,
      tag,
    } = this.props;
    const { ordering, tagFilters, sortMenuAnchorElement } = this.state;

    const searchPlaceholder = intl.formatMessage({
      id: 'app.program_collection.search',
      description: 'Placeholder for search entry',
      defaultMessage: 'Search...',
    });

    const tagFilterPlaceholder = intl.formatMessage({
      id: 'app.program_collection.filter',
      description: 'Placeholder for selecting tags used to filter',
      defaultMessage: 'Tag filters...',
    });

    const sortText = intl.formatMessage({
      id: 'app.program_collection.sort',
      description: 'Button label for sort options',
      defaultMessage: 'Sort',
    });

    const Title = withStyles(() => ({
      root: {
        fontWeight: 'bold',
      },
    }))(Typography);

    const FixedWidthAutocomplete = withStyles(() => ({
      root: {
        width: '100%',
      },
    }))(Autocomplete);

    const TitleArea = withStyles((theme) => ({
      root: {
        marginBottom: theme.spacing(4),
      },
    }))(Grid);

    const PaddedBox = withStyles((theme) => ({
      root: {
        marginTop: theme.spacing(3),
        marginBottom: theme.spacing(3),
      },
    }))(Box);

    return (
      <Container maxWidth="lg">
        <TitleArea item container direction="row" justify="space-between">
          <Grid item>
            <Title variant="h4">
              {label}
            </Title>
          </Grid>
          <Grid item>
            <Button
              variant="contained"
              color="primary"
              size="large"
              startIcon={<Add />}
              component={Link}
              to="/mission-control"
            >
              <FormattedMessage
                id="app.program_list.new"
                description="Button label to create new program"
                defaultMessage="New Program"
              />
            </Button>
          </Grid>
        </TitleArea>
        <Grid container spacing={2}>
          <Grid item>
            <TextField
              variant="outlined"
              size="small"
              className="prompt"
              icon="search"
              placeholder={searchPlaceholder}
              onChange={this.handleSearchChange}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <Search />
                  </InputAdornment>
                ),
              }}
            />
          </Grid>
          <Grid item>
            <Button aria-controls="sort-menu" aria-haspopup="true" onClick={this.handleSortClick}>
              {sortText}
            </Button>
            <Menu
              id="sort-menu"
              anchorEl={sortMenuAnchorElement}
              keepMounted
              open={Boolean(sortMenuAnchorElement)}
              onClose={this.handleSortClose}
              getContentAnchorEl={null}
              anchorOrigin={{
                vertical: 'bottom',
                horizontal: 'center',
              }}
              transformOrigin={{
                vertical: 'top',
                horizontal: 'center',
              }}
            >
              <MenuItem onClick={this.handleNameOrderingChange}>
                <FormattedMessage
                  id="app.program_collection.name"
                  description="Button label to sort by name"
                  defaultMessage="Name"
                />
                {
                  ordering === 'name' ? (
                    <>
                      <ArrowDownward />
                      a ... z
                    </>
                  ) : (
                    <>
                      <ArrowUpward />
                      z ... a
                    </>
                  )
                }
              </MenuItem>
            </Menu>
          </Grid>
          <Grid item xs={2}>
            <FixedWidthAutocomplete
              id="tag-select"
              multiple
              freeSolo
              filterSelectedOptions
              size="small"
              options={tag.tags.map((t) => t.name)}
              onChange={this.handleTagFilterChange}
              value={tagFilters}
              renderInput={(params) => (
                <TextField {...params} variant="outlined" label={tagFilterPlaceholder} />
              )}
              renderTags={(value, getTagProps) => value.map((option, index) => (
                <Chip color="secondary" size="small" label={option} {...getTagProps({ index })} />
              ))}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3}>
          {
            programs.results.map((program) => (
              <Grid item xs={12} md={6} lg={3}>
                <Card key={program.id}>
                  <CardHeader
                    title={program.name}
                    subheader={
                        user.username === program.user.username ? (
                          <FormattedMessage
                            id="app.program_collection.mine"
                            description="Label to indicate program owned by user"
                            defaultMessage="Mine"
                          />
                        ) : program.user.username
                      }
                  />
                  {/* <CardActionArea>
                    <CardContent>
                      Lorem ipsum dolor sit amet, consectetur adipiscing elit,
                      sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                    </CardContent>
                  </CardActionArea> */}
                  <CardActions>
                    <Button
                      id={program.id}
                      data-owned={user.username === program.user.username}
                      onClick={onProgramClick}
                    >
                      {
                        user.username === program.user.username ? (
                          <FormattedMessage
                            id="app.program_collection.work"
                            description="Button label to keep working on program"
                            defaultMessage="Keep Working"
                          />
                        ) : (
                          <FormattedMessage
                            id="app.program_collection.view"
                            description="Button label to view a program"
                            defaultMessage="View"
                          />
                        )
                      }
                    </Button>
                    {
                      user.username === program.user.username ? (
                        <Button
                          id={program.id}
                          name={program.name}
                          onClick={onRemoveClick}
                          floated="right"
                        >
                          <FormattedMessage
                            id="app.program_collection.remove"
                            description="Button label to remove a program"
                            defaultMessage="Remove"
                          />
                        </Button>
                      ) : (null)
                    }
                  </CardActions>
                </Card>
              </Grid>
            ))
          }
        </Grid>
        {
          programs.total_pages > 1 ? (
            <PaddedBox
              display="flex"
              justifyContent="center"
            >
              <Pagination
                defaultPage={1}
                count={programs.total_pages}
                showFirstButton
                showLastButton
                color="secondary"
                onChange={this.handlePageChange}
              />
            </PaddedBox>
          ) : (null)
        }
      </Container>
    );
  }
}

ProgramCollection.defaultProps = {
  tag: {
    tags: [],
  },
};

ProgramCollection.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }).isRequired,
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  label: PropTypes.string.isRequired,
  onProgramClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(ProgramCollection));
