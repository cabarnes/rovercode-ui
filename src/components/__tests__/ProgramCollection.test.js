import React from 'react';
import {
  Button,
  Card,
  Dropdown,
  Header,
  Input,
} from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import CustomPagination from '../CustomPagination';
import ProgramCollection from '../ProgramCollection';

let onProgramClick;
let onRemoveClick;
let onUpdate;
let adminUser;
let testUser;

describe('The ProgramCollection component', () => {
  beforeEach(() => {
    onProgramClick = jest.fn();
    onRemoveClick = jest.fn();
    onUpdate = jest.fn();
    adminUser = { username: 'admin' };
    testUser = { username: 'testuser' };
  });

  test('renders on the page with no errors', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('shows the correct number of programs for the user', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={testUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Card).length).toBe(2);
    expect(wrapper.find(Card.Meta).first().children().find(FormattedMessage)
      .prop('defaultMessage')).toBe('Mine');
    expect(wrapper.find(Button).first().children().find(FormattedMessage)
      .prop('defaultMessage')).toBe('Keep Working');
  });

  test('shows the correct number of programs for other users', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Card).length).toBe(2);
    expect(wrapper.find(Card.Meta).first().prop('children')).toBe('testuser');
    expect(wrapper.find(Button).first().children().find(FormattedMessage)
      .prop('defaultMessage')).toBe('View');
  });

  test('callback when program click', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find(Button).first().simulate('click', {
      target: {
        id: 33,
      },
    });

    expect(onProgramClick).toHaveBeenCalledWith({
      target: {
        id: 33,
      },
    });
  });

  test('callback when remove click', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find(Button).last().simulate('click', {
      target: {
        id: 33,
      },
    });

    expect(onRemoveClick).toHaveBeenCalledWith({
      target: {
        id: 33,
      },
    });
  });

  test('callback when page changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find(CustomPagination).simulate('pageChange', null, {
      activePage: 2,
    });

    expect(onUpdate).toHaveBeenCalledWith({
      page: 2,
      ordering: 'name',
      tag: '',
    });
  });

  test('callback when search changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find(Input).simulate('change', {
      target: {
        value: 'abc',
      },
    });

    expect(onUpdate).toHaveBeenCalledWith({
      search: 'abc',
      page: 1,
      ordering: 'name',
      tag: '',
    });
  });

  test('callback when order changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find({ name: 'name' }).simulate('click', null, {
      name: 'name',
    });

    expect(onUpdate).toHaveBeenCalledWith({
      page: 1,
      ordering: '-name',
      tag: '',
    });

    wrapper.find({ name: 'name' }).simulate('click', null, {
      name: 'name',
    });

    expect(onUpdate).toHaveBeenCalledWith({
      page: 1,
      ordering: 'name',
      tag: '',
    });

    wrapper.instance().handleOrderingChange(null, {
      name: 'field_name',
    });

    expect(onUpdate).toHaveBeenCalledWith({
      page: 1,
      ordering: 'field_name',
      tag: '',
    });
  });

  test('callback when tag changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };

    const tag = {
      tags: [{
        name: 'tag1',
      }],
    };

    const wrapper = shallowWithIntl(
      <ProgramCollection
        programs={programs}
        tag={tag}
        label="My Programs"
        user={adminUser}
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onUpdate={onUpdate}
      />,
    ).dive().dive();

    wrapper.find(Dropdown).last().simulate('change', {}, {
      value: [
        'tag1',
        'tag2',
      ],
    });

    expect(onUpdate).toHaveBeenCalledWith({
      page: 1,
      ordering: 'name',
      tag: 'tag1,tag2',
    });
  });
});
