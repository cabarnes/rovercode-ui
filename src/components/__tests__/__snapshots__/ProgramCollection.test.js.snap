// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramCollection component renders on the page with no errors 1`] = `
<Fragment>
  <Grid>
    <GridRow>
      <GridColumn
        width={2}
      />
      <GridColumn
        width={4}
      >
        <Input
          className="prompt"
          icon="search"
          onChange={[Function]}
          placeholder="Search..."
          type="text"
        />
      </GridColumn>
      <GridColumn
        width={4}
      >
        <Header
          as="h1"
          textAlign="center"
        >
          My Programs
        </Header>
      </GridColumn>
      <GridColumn
        width={1}
      />
      <GridColumn
        floated="right"
        width={2}
      >
        <Dropdown
          additionLabel="Add "
          additionPosition="top"
          button={true}
          className="icon"
          closeOnBlur={true}
          closeOnEscape={true}
          deburr={false}
          floating={true}
          icon="sort"
          labeled={true}
          minCharacters={1}
          noResultsMessage="No results found."
          openOnFocus={true}
          renderLabel={[Function]}
          searchInput="text"
          selectOnBlur={true}
          selectOnNavigation={true}
          text="Sort"
          wrapSelection={true}
        >
          <DropdownMenu>
            <DropdownItem
              name="name"
              onClick={[Function]}
            >
              <FormattedMessage
                defaultMessage="Name"
                id="app.program_collection.name"
                values={Object {}}
              />
              <Icon
                as="i"
                name="caret down"
              />
            </DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </GridColumn>
      <GridColumn
        width={3}
      >
        <Dropdown
          additionLabel="Add "
          additionPosition="top"
          closeOnBlur={true}
          closeOnEscape={true}
          deburr={false}
          icon="dropdown"
          minCharacters={1}
          multiple={true}
          noResultsMessage="No results found."
          onChange={[Function]}
          openOnFocus={true}
          options={Array []}
          placeholder="Tag filters..."
          renderLabel={[Function]}
          scrolling={true}
          searchInput="text"
          selectOnBlur={true}
          selectOnNavigation={true}
          selection={true}
          value={Array []}
          wrapSelection={true}
        />
      </GridColumn>
    </GridRow>
  </Grid>
  <CardGroup
    centered={true}
  >
    <Card
      key="33"
    >
      <CardContent>
        <CardHeader>
          Unnamed_Design_3
        </CardHeader>
        <CardMeta>
          <FormattedMessage
            defaultMessage="Mine"
            id="app.program_collection.mine"
            values={Object {}}
          />
        </CardMeta>
      </CardContent>
      <CardContent
        extra={true}
      >
        <Button
          as="button"
          data-owned={true}
          id={33}
          onClick={[MockFunction]}
          primary={true}
        >
          <FormattedMessage
            defaultMessage="Keep Working"
            id="app.program_collection.work"
            values={Object {}}
          />
        </Button>
        <Button
          as="button"
          floated="right"
          id={33}
          name="Unnamed_Design_3"
          negative={true}
          onClick={[MockFunction]}
        >
          <FormattedMessage
            defaultMessage="Remove"
            id="app.program_collection.remove"
            values={Object {}}
          />
        </Button>
      </CardContent>
    </Card>
    <Card
      key="5"
    >
      <CardContent>
        <CardHeader>
          Unnamed_Design_2
        </CardHeader>
        <CardMeta>
          testuser
        </CardMeta>
      </CardContent>
      <CardContent
        extra={true}
      >
        <Button
          as="button"
          data-owned={false}
          id={5}
          onClick={[MockFunction]}
          primary={true}
        >
          <FormattedMessage
            defaultMessage="View"
            id="app.program_collection.view"
            values={Object {}}
          />
        </Button>
      </CardContent>
    </Card>
  </CardGroup>
  <Grid
    centered={true}
  >
    <GridRow>
      <WithStyles(ForwardRef(Pagination))
        color="secondary"
        count={2}
        defaultPage={1}
        onChange={[Function]}
        showFirstButton={true}
        showLastButton={true}
      />
    </GridRow>
  </Grid>
</Fragment>
`;
